File: src/lib/likes.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';

const KEY_PREFIX = 'likes:';
const LOCAL_FILE = join(process.cwd(), '.data', 'likes.json');

export function likesKey(id: string): string {
  return `${KEY_PREFIX}${id}`;
}

export function isPostId(id: string): boolean {
  return /^[a-z0-9][a-z0-9/_-]{0,199}$/i.test(id);
}

function readEnv(name: string): string | undefined {
  const raw = process.env[name];
  if (typeof raw !== 'string') return undefined;
  const value = raw.trim().replace(/^['"]|['"]$/g, '');
  return value || undefined;
}

function redisConfig() {
  const url = readEnv('UPSTASH_REDIS_REST_URL') ?? readEnv('KV_REST_API_URL');
  const token = readEnv('UPSTASH_REDIS_REST_TOKEN') ?? readEnv('KV_REST_API_TOKEN');
  if (!url || !token) return null;
  return { url, token };
}

async function redis(command: Array<string | number>): Promise<unknown> {
  const config = redisConfig();
  if (!config) throw new Error('Redis is not configured');

  const response = await fetch(config.url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${config.token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(command),
  });

  if (!response.ok) {
    throw new Error(`Redis ${response.status}`);
  }

  const payload = (await response.json()) as { result?: unknown; error?: string };
  if (payload.error) throw new Error(payload.error);
  return payload.result;
}

async function readLocal(): Promise<Record<string, number>> {
  try {
    return JSON.parse(await readFile(LOCAL_FILE, 'utf8')) as Record<string, number>;
  } catch {
    return {};
  }
}

async function writeLocal(data: Record<string, number>): Promise<void> {
  await mkdir(dirname(LOCAL_FILE), { recursive: true });
  await writeFile(LOCAL_FILE, `${JSON.stringify(data, null, 2)}\n`);
}

function toCount(value: unknown): number {
  const n = typeof value === 'number' ? value : Number(value ?? 0);
  return Number.isFinite(n) && n > 0 ? Math.floor(n) : 0;
}

export async function getLikes(id: string): Promise<number> {
  if (redisConfig()) {
    return toCount(await redis(['GET', likesKey(id)]));
  }
  if (import.meta.env.DEV) {
    const data = await readLocal();
    return toCount(data[id]);
  }
  throw new Error('Likes store is not configured');
}

export async function bumpLikes(id: string, delta: 1 | -1): Promise<number> {
  if (redisConfig()) {
    const raw = await redis(['INCRBY', likesKey(id), delta]);
    const n = typeof raw === 'number' ? raw : Number(raw ?? 0);
    if (!Number.isFinite(n) || n <= 0) {
      await redis(['SET', likesKey(id), 0]);
      return 0;
    }
    return Math.floor(n);
  }
  if (import.meta.env.DEV) {
    const data = await readLocal();
    const next = Math.max(0, toCount(data[id]) + delta);
    data[id] = next;
    await writeLocal(data);
    return next;
  }
  throw new Error('Likes store is not configured');
}

File: api/likes.go
package handler

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
)

const keyPrefix = "likes:"

var (
	postIDRe   = regexp.MustCompile(`(?i)^[a-z0-9][a-z0-9/_-]{0,199}$`)
	localMu    sync.Mutex
	errNoStore = errors.New("not_configured")
)

type countBody struct {
	Count int `json:"count"`
}

type errorBody struct {
	Error  string `json:"error"`
	Reason string `json:"reason,omitempty"`
}

type actionBody struct {
	Action string `json:"action"`
}

func Handler(w http.ResponseWriter, r *http.Request) {
	id := postID(r)
	if !postIDRe.MatchString(id) {
		writeJSON(w, http.StatusBadRequest, errorBody{Error: "Invalid post"})
		return
	}

	switch r.Method {
	case http.MethodGet:
		n, err := getLikes(id)
		if err != nil {
			writeStoreError(w, err)
			return
		}
		writeJSON(w, http.StatusOK, countBody{Count: n})
	case http.MethodPost:
		var body actionBody
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, http.StatusBadRequest, errorBody{Error: "Invalid body"})
			return
		}
		var delta int64
		switch body.Action {
		case "like":
			delta = 1
		case "unlike":
			delta = -1
		default:
			writeJSON(w, http.StatusBadRequest, errorBody{Error: "Invalid action"})
			return
		}
		n, err := bumpLikes(id, delta)
		if err != nil {
			writeStoreError(w, err)
			return
		}
		writeJSON(w, http.StatusOK, countBody{Count: n})
	default:
		w.Header().Set("Allow", "GET, POST")
		writeJSON(w, http.StatusMethodNotAllowed, errorBody{Error: "Method not allowed"})
	}
}

func postID(r *http.Request) string {
	if id := strings.TrimSpace(r.URL.Query().Get("id")); id != "" {
		return id
	}
	path := strings.Trim(r.URL.Path, "/")
	const prefix = "api/likes/"
	if strings.HasPrefix(path, prefix) {
		return strings.Trim(strings.TrimPrefix(path, prefix), "/")
	}
	return ""
}

func writeStoreError(w http.ResponseWriter, err error) {
	reason := "store_failed"
	if errors.Is(err, errNoStore) {
		reason = "not_configured"
	}
	writeJSON(w, http.StatusServiceUnavailable, errorBody{
		Error:  "Likes are unavailable",
		Reason: reason,
	})
}

func writeJSON(w http.ResponseWriter, status int, data any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(data)
}

func env(name string) string {
	v := strings.TrimSpace(os.Getenv(name))
	return strings.Trim(v, `"'`)
}

func redisConfig() (url, token string, ok bool) {
	url = env("UPSTASH_REDIS_REST_URL")
	if url == "" {
		url = env("KV_REST_API_URL")
	}
	token = env("UPSTASH_REDIS_REST_TOKEN")
	if token == "" {
		token = env("KV_REST_API_TOKEN")
	}
	return url, token, url != "" && token != ""
}

func onVercel() bool {
	return env("VERCEL") != ""
}

func redis(command []string) (any, error) {
	url, token, ok := redisConfig()
	if !ok {
		return nil, errNoStore
	}
	payload, err := json.Marshal(command)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return nil, errors.New("redis " + strconv.Itoa(res.StatusCode))
	}
	var out struct {
		Result any    `json:"result"`
		Error  string `json:"error"`
	}
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	if out.Error != "" {
		return nil, errors.New(out.Error)
	}
	return out.Result, nil
}

func toCount(v any) int {
	n, ok := asInt64(v)
	if !ok || n <= 0 {
		return 0
	}
	return int(n)
}

func getLikes(id string) (int, error) {
	if _, _, ok := redisConfig(); ok {
		v, err := redis([]string{"GET", keyPrefix + id})
		if err != nil {
			return 0, err
		}
		return toCount(v), nil
	}
	if onVercel() {
		return 0, errNoStore
	}
	data, err := readLocal()
	if err != nil {
		return 0, err
	}
	return toCount(data[id]), nil
}

func bumpLikes(id string, delta int64) (int, error) {
	if _, _, ok := redisConfig(); ok {
		v, err := redis([]string{"INCRBY", keyPrefix + id, strconv.FormatInt(delta, 10)})
		if err != nil {
			return 0, err
		}
		n := int64(toCount(v))
		if raw, ok := asInt64(v); ok {
			n = raw
		}
		if n < 0 {
			if _, err := redis([]string{"SET", keyPrefix + id, "0"}); err != nil {
				return 0, err
			}
			return 0, nil
		}
		return int(n), nil
	}
	if onVercel() {
		return 0, errNoStore
	}
	localMu.Lock()
	defer localMu.Unlock()
	data, err := readLocal()
	if err != nil {
		return 0, err
	}
	next := toCount(data[id]) + int(delta)
	if next < 0 {
		next = 0
	}
	data[id] = next
	if err := writeLocal(data); err != nil {
		return 0, err
	}
	return next, nil
}

func asInt64(v any) (int64, bool) {
	switch n := v.(type) {
	case nil:
		return 0, true
	case int:
		return int64(n), true
	case int64:
		return n, true
	case float64:
		return int64(n), true
	case json.Number:
		i, err := n.Int64()
		return i, err == nil
	case string:
		if n == "" {
			return 0, true
		}
		i, err := strconv.ParseInt(n, 10, 64)
		return i, err == nil
	default:
		return 0, false
	}
}

func localFile() string {
	return filepath.Join(".", ".data", "likes.json")
}

func readLocal() (map[string]int, error) {
	raw, err := os.ReadFile(localFile())
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]int{}, nil
		}
		return nil, err
	}
	var data map[string]int
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	if data == nil {
		data = map[string]int{}
	}
	return data, nil
}

func writeLocal(data map[string]int) error {
	if err := os.MkdirAll(filepath.Dir(localFile()), 0o755); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(localFile(), raw, 0o644)
}
